\documentclass[11pt]{article}
\pdfoutput=1
\usepackage{simplemargins}
\usepackage[pdftex]{graphicx}
\usepackage{setspace}
\graphicspath{{figures/}}
\usepackage{siunitx}
\setlength{\parindent}{0pt} 
\setlength{\parskip}{1.6ex} 
\setallmargins{1in} 
\linespread{1.6}
\usepackage[round]{natbib}
\usepackage{color}
\usepackage{subfigure}
\usepackage{booktabs}
\usepackage{pdflscape}
\usepackage{tabularx}
\usepackage[colorlinks=false]{hyperref}
\usepackage[anythingbreaks]{breakurl}
\usepackage{longtable}
\usepackage[tableposition=top]{caption}
\listfiles

% make subfigure labels capitalized
\renewcommand{\thesubfigure}{(\Alph{subfigure})}

%setup supplement
\newcommand{\beginsupplement}{%
        \setcounter{table}{0}
        \renewcommand{\thetable}{S\arabic{table}}
        \setcounter{figure}{0}
        \renewcommand{\thefigure}{S\arabic{figure}}
        \renewcommand{\thesection}{S\arabic{section}}
        \renewcommand{\thesubsection}{S\arabic{subsection}} 
     }

\begin{document}
\newdimen\LTcapwidth \LTcapwidth=\textwidth


% title must be 150 characters or less
% \begin{flushleft} 
% \singlespacing
% {\large \textbf{title}}

% insert author names, affiliations and corresponding author email.
% author1$^{1}$, 
% author2$^{2}$,


% \bf{1} affil 1
% \\
% \bf{2} affil 2
% \\

% $\ast$ e-mail: corresponding\_author@domain.com
% \end{flushleft}

% \section*{Abstract}

% \section*{Introduction}

\section*{Materials and methods}\label{ss:mats}
A total of 7232 chromatogram files from two sets of replicate samples,
two each for needle and cambium were processed using an IPython
\citep[v 2.1]{per-gra:2007} notebook and various utilities.  The
chromatograms were processed using Phred \citep[v.\
020425]{ewing1998b} into Fasta sequences with a trimming cutoff
probability of 0.01 (e.g., Phred score = 20), and sequences which were
less than 100 bases long were excluded.  The ESTs from all replicates
were combined into single file (where all sequence identifiers
indicated the source and replicate of the EST).  This combined file
was process first using SeqClean \citep[v.\ 2/2011][default
parameters]{citeulike:1911083}, screening both for vectors using the
UniVec \citep{cochrane01012010} database and contamination from
\textit{E. coli} K-12 substr.\ DH10B.  The cleaned and trimmed ESTs
were in each case were subsequently assembled into unigenes using
iAssembler \citep[v.\ 1.3.2][default parameters]{zheng2011iassembler}.

The assembled unigenes were aligned to a local copy of the nr database
(downloaded in February 2014) with BLASTX \citep{citeulike:238188}
keeping, at most, the top 10 hits with an e-value cutoff of
$e^{-5}$. The resulting alignment XML file were filtered using
BioPython \citep[v.\ 1.6.4]{citeulike:4202607} to include only the top
HSPs (i.e., alignments) which were at least 30\% of both unigene
length and identity as well as only those hits which aligned to a
sequence assigned to the NCBI taxonomic division, ``Plants''. These
filtered BLAST results were imported into Blast2GO Pro \citep[v.\
2.7.2, bg2\_sep14]{citeulike:2733895} for functional annotation using
the default parameters (e.g., keeping at most 10 hits with an HSP
length cutoff of 33 amino acids).  Annotations were assigned using
default parameters (e.g., e-value: $e^{-6}$, annotation cutoff: 55, go
weight: 5, HSP-hit coverage cutoff: 0, taxonomy filtering: none) and
evidence code weights. The full suite of InterPro
\citep{citeulike:12942060} mappings were performed for each unigene
with a valid BLAST result and these results were merged with the GO
annotations.  These annotations were augmented with ANNEX
\citep{annex} and mapped to enzyme codes and KEGG
\citep{citeulike:9172127} pathways. The resulting annotations were
then exported to a text file for further analysis. 

Differential gene (unigene) expression (DGE) was computed in R
\citep[v. 3.1.1]{R} using RSEM
\citep[v. 1.2.15]{Li:BmcBioinformatics:2011} in the standard way using
bowtie \citep[v. 2.2.3]{Langmead:NatMethods:2012} by first creating a
reference (\texttt{rsem-prepare-reference}) and then the computing the
expression on both a parent-wise and tissue-wise basis
(\texttt{rsem-calculate-expression}). Models were evaluated using
\texttt{rsem-plot-model} and the parent-wise gene results were
combined for further analysis
(\texttt{rsem-generate-data-matrix}). The resulting data matrix was
used in an EBSeq \citep{Leng:Bioinformatics:2013} pipeline to compute
differential gene expression (50 rounds, accounting for differences in
library size using median normalization), considering all unigenes
($Qtrm=1$) to which at least five cleaned ESTs mapped
($QtrmCut=5$). Multiple testing was controlled by FDR
\citep{benjamini1995controlling} at 0.05 using
\texttt{rsem-control-fdr}.  As an additional check, convergence of the
EBTest procedure was confirmed by examining the parameters Alpha,
Beta, and P individually.  These values converged at 33, 32, and 32
generations, respectively, providing confidence in the estimates for
differential expression. The count matrix used for DGE, in concert
with the exported annotations from Blast2GO was also used to compute
GO enrichment for each ontology (i.e., BP, CC, MF), using topGO
\citep{topgo}, by tissue type, for nodes in the GO ontologies
containing at least 5 genes.  Tests for enrichment in GO terms 
were computed using Fisher's exact test and controlled using FDR.

All relevant code for both the analysis and manuscript can be found at
\url{http://www.github.com/cfriedline/black_spruce}.  Raw data files
can be accessed from the iPlant Data Store at
\url{https://de.iplantc.org/de/?type=data&folder=/iplant/home/cfriedline/pub_data/black_spruce}
and the quality-controlled ESTs can be found in NCBI dbEST under accessions
XXX--XXX.

\section*{Results}

A total of 7232 chromatogram files were obtained from sequencing for
two biological replicates (e.g., P32, P40) for two tissue types:
needle and cambium.  Following processing with Phred and filtering by
length, 5996 raw ESTs were combined into a single file for downstream
analysis.  After processing with Seqclean, 5938 ESTs remained; 2842
were trimmed and 58 were removed from the dataset, either by mapping
to \textit{E. coli} (34), low complexity (1), or length/shortq (23).
Assembly resulted in 1945 unigenes with an average coverage of
$\approx{3}$ ESTs ($\pm \approx{10}, range=[1,274]$). The EST assbemly 
is summarized in Table \ref{tab:est}.

DGE was evaluated for 169 unigenes (transcripts) meeting the cutoff
and 35 of these were reported as significant at FDR = 0.05 (Figure
\ref{fig:unigene_venn}).  In cambium, 23 of these unigenes were
differentially expressed as were 12 in the needle tissue, though while
20 unigenes were uniquely expressed in cambium, there were none
uniquely expressed in the needles.  The BP annotations for these
unigenes are outlined in Table \ref{tab:dge_bp}.

There were 933 unigenes ($\sim 48\%$) with GO annotations (mean=$6 \pm
5$ annotations/unigene, range=[1, 44]). Of these unigenes with GO
annotations, 767, 600, 800 were usable by the topGO enrichment tests
for the BP, CC, and MF ontologies, respectively.  The top 20 terms for
each ontology which were signficantly ($q < 0.05$) enriched for each
ontology are displayed in Figure \ref{fig:go_full_combined_20_bh}.  In
addition to the overall summarization of the significantly enriched GO
terms, Figure \ref{fig:go_full_20_bh} shows the top 20 terms (at most) 
which were signficantly enriched, by tissue type.


% \section*{Discussion}

% \section*{Acknowledgements}

\clearpage

\singlespacing
\bibliographystyle{spbasic}
\bibliography{refs}

\clearpage
%figures

\begin{figure}[t]
  \centering
  \includegraphics[width=\textwidth,keepaspectratio]{unigene_venn.pdf}
  \caption{Number expressed unigenes (top) across tissues as well as
    differential expression patterns (bottom). Of the 15 shared
    differnetially expressed genes, 3 were differntial in cambium and
    12 were differntially expressed in needle.}
  \label{fig:go_venn}
\end{figure}



\begin{figure}[t]
  \centering
  \includegraphics[height=8.5in,keepaspectratio]{go_full_combined_20_bh}
  \caption{Top 20 Significantly enriched ($q < 0.05$) GO terms for
    each ontology. The total number of significant terms for each
    ontology is shown in parentheses.}
  \label{fig:go_full_combined_20_bh}
\end{figure}


\begin{figure}[t]
  \centering
  \includegraphics[width=\textwidth]{go_full_20_bh}
  \caption{Top 20 Significantly enriched ($q < 0.05$) GO terms for
    each ontology and tissue type. The total number of significant
    terms for each ontology and tissue type is shown in parentheses}
  \label{fig:go_full_20_bh}
\end{figure}

\clearpage

%tables

\input{est_table}
\input{dge_bp_mod}

\clearpage

\beginsupplement

\section*{Supplementary material}

% %put supplementary figures and tables below here
\subsection*{Supplementary Figures}\label{ss:supp-fig}

\begin{figure}[t]
  \centering
  \includegraphics[height=8.5in,keepaspectratio]{go_full_combined_20}
  \caption{Top 20 Significantly enriched ($p < 0.05$) GO terms for
    each ontology. The total number of significant terms for each
    ontology is shown in parentheses.}
  \label{fig:go_combined}
\end{figure}


\begin{figure}[t]
  \centering
  \includegraphics[width=\textwidth]{go_full_20}
  \caption{Top 20 Significantly enriched ($p < 0.05$) GO terms for
    each ontology and tissue type. The total number of significant
    terms for each ontology and tissue type is shown in parentheses}
  \label{fig:go_tissue}
\end{figure}
 
\clearpage

\subsection*{Supplementary Tables}\label{ss:supp-tab}
\scriptsize
\input{go_cambium_BP_formatted}  
\clearpage
\input{go_cambium_MF_formatted}
\clearpage
\input{go_cambium_CC_formatted}
\clearpage
\input{go_needle_BP_formatted}  
\clearpage
\input{go_needle_MF_formatted}
\clearpage
\input{go_needle_CC_formatted}


\end{document}